File: electron/ipc/vibe-file.test.ts
import {it,expect,vi} from 'vitest';
import axios from 'axios';
import {prepareExtras,buildPayload,countCachedVibes} from './nai';
import {DEFAULT_PARAMS} from '../../src/types';
const model='nai-diffusion-4-5-full';
it('uses imported encodings verbatim, without image decoding or paid encode requests',async()=>{
 const spy=vi.spyOn(axios,'post').mockRejectedValue(new Error('Unexpected network'));
 try {
  const extras={vibeImages:[{base64:'',encodings:[{model,infoExtracted:.7,encoding:'AQIDBA=='}],infoExtracted:.7,strength:.31}],charCaptions:[],preciseReferences:[]};
  expect(countCachedVibes(extras,{...DEFAULT_PARAMS,model})).toBe(1);
  const prepared=await prepareExtras({...DEFAULT_PARAMS,model},extras);
  const payload=buildPayload({...DEFAULT_PARAMS,model},123,prepared);
  expect(payload.parameters.reference_image_multiple).toEqual(['AQIDBA==']);
  expect(payload.parameters.reference_strength_multiple).toEqual([.31]);
  expect(spy).not.toHaveBeenCalled();
  await expect(prepareExtras({...DEFAULT_PARAMS,model:'nai-diffusion-4-5-curated'},extras)).rejects.toThrow('no original image');
  await expect(prepareExtras({...DEFAULT_PARAMS,model:'nai-diffusion-5-full'},extras)).rejects.toThrow('Vibe');
  expect(spy).not.toHaveBeenCalled();
 } finally {spy.mockRestore();}
});

File: mobile/test/vibe_file_test.dart
import 'dart:convert';
import 'dart:io';
import 'package:flutter_test/flutter_test.dart';
import 'package:novelai_mobile/models/nai_models.dart';
import 'package:novelai_mobile/services/vibe_file.dart';
import 'package:novelai_mobile/services/nai_api.dart';

void main() {
  const model = 'nai-diffusion-4-5-full';
  Map<String, dynamic> entry() => {
        'identifier': 'novelai-vibe-transfer',
        'version': 1,
        'type': 'encoding',
        'encodings': {
          'v4-5full': {
            'unknown': {'encoding': 'AQIDBA=='}
          }
        },
        'importInfo': {
          'model': model,
          'information_extracted': .7,
          'strength': .31
        }
      };
  test('encoding-only import, copies and export retain model and extraction',
      () {
    final refs = parseVibeFile(jsonEncode(entry()));
    expect(refs.single.base64, '');
    expect(refs.single.infoExtracted, .7);
    expect(refs.single.strength, .31);
    expect(refs.single.matchingEncoding(model), 'AQIDBA==');
    expect(
        GenerateExtras(vibeImages: refs)
            .copy()
            .vibeImages
            .single
            .matchingEncoding(model),
        'AQIDBA==');
    expect(
        VibeTransferItem.fromJson(refs.single.copyWith(strength: .4).toJson())
            .matchingEncoding(model),
        'AQIDBA==');
    expect(parseVibeFile(exportVibeFile(refs)).single.toJson(),
        refs.single.toJson());
    expect(() => refs.single.validateModel('nai-diffusion-4-5-curated'),
        throwsStateError);
  });
  test('builds payload with imported encoding and no paid network request',
      () async {
    final api = NaiApi(),
        extras = GenerateExtras(vibeImages: parseVibeFile(jsonEncode(entry())));
    expect(api.countCachedVibes(model, extras), 1);
    final result = await api.buildPayload(
        'unused',
        AppSettings(proxyMode: 'direct'),
        GenerateParams(model: model),
        123,
        extras);
    final p = result['parameters'] as Map;
    expect(p['reference_image_multiple'], ['AQIDBA==']);
    expect(p['reference_strength_multiple'], [.31]);
  });
  test('invalid bundles fail before changing state', () {
    for (final v in [
      {},
      {...entry(), 'version': 2},
      {...entry(), 'encodings': {}},
      {
        ...entry(),
        'importInfo': {'information_extracted': 2}
      },
      {'identifier': 'novelai-vibe-transfer-bundle', 'version': 1, 'vibes': []}
    ]) {
      expect(() => parseVibeFile(jsonEncode(v)), throwsFormatException);
    }
  });
  test('provided bundle uses both encodings verbatim', () async {
    final path = Platform.environment['VIBE_FIXTURE'];
    if (path == null) return;
    final refs = parseVibeFile(await File(path).readAsString());
    expect(refs.length, 2);
    expect(refs.map((v) => v.strength), [.31, .24]);
    expect(refs.map((v) => v.infoExtracted), [.7, .44]);
    final p = await NaiApi().buildPayload(
        'unused',
        AppSettings(proxyMode: 'direct'),
        GenerateParams(model: model),
        123,
        GenerateExtras(vibeImages: refs));
    expect((p['parameters'] as Map)['reference_image_multiple'],
        refs.map((r) => r.matchingEncoding(model)).toList());
  });
}
